### ISSUE TITLE
noise(x, y) always return 0.46875

### ISSUE THREAD

**Issue by bofeng** [processing-js/processing-js#241]
The java version works well, but the js version, everytime call noise(x, y), it returns 0.46875

```
for (int x = 0; x < 5; x++) {
    for (int y = 0; y < 5; y++) {
        n = noise(x, y);
        text("number: " + n, 10, 10 * i);
        i = i + 1;
    }
}
```

Run it here : http://sketchpad.cc/sp/pad/view/mPI5P5i6so/latest

**Comment by teo1978** [comment]
Looks like it has to do with `x` and `y` being integer. If you add `0.99` instead of 1 in each iteration, the resulting values do change.

**Comment by bofeng** [comment]
@teo1978 Thanks, it is wired though, still a bug ... besides even change it to ```noise(x - 0.01, y - 0.01)```, the number it generated is still around 0.46875, while the Java version is much more random

![processing-bug](https://cloud.githubusercontent.com/assets/3884670/15659561/9297693c-269a-11e6-9ff3-b3f0279cbb58.png)

**Comment by teo1978** [comment]
> still a bug

Of course! I have no idea whether the values generated by processing.js make some sense (I mean, if they are valid Perlin noise according to a different "scale" or whatever) or if it's completely broken. But either way, as long as it does not generate *exactly* the same values as the Java version for any given input and seed, it's of course a bug.

**Comment by Pomax** [comment]
They *should* make sense, since they're proper Perlin noise, but it's possible that there are subtle bugs in the implementation on https://github.com/processing-js/processing-js/blob/c2b14664eae6d0f036e68e69f1f32afcbde88f5f/src/P5Functions/Math.js#L612

**Comment by GoToLoop** [comment]
The fact that Pjs' **noise()** doesn't match Java Mode is an old knowledge.
In this online sketch link here: http://studio.ProcessingTogether.com/sp/pad/export/ro.9Ql2E8JBC54LJ
The brown mountain `^` rarely shows up. However, if we paste the code in Java Mode, all elements' distribution within the "world" there are what is expected.

**Comment by Pomax** [comment]
it's not old knowledge unless it's documented where people download processing.js, which I'm pretty sure is not the case.

**Comment by GoToLoop** [comment]
More like that bug has showed up in the Processing forum a few times:

1. https://forum.Processing.org/two/discussion/15773/noise-function-on-webpages
2. https://forum.Processing.org/two/discussion/6838/1d-array-affecting-2d-array#Item_4
3. https://forum.Processing.org/two/discussion/8858/does-noiseseed-work-on-javascript-mode

And even some tidbits in this very place here: https://github.com/processing-js/processing-js/issues/131

Here's another online sketch example: http://studio.ProcessingTogether.com/sp/pad/export/ro.91FXxn8BG2fPL/latest

Running under Pjs, all "walkers" almost always go towards the top left only.
Pasting it in Java Mode, although it tends to go to top-left too, it's totally much more random & balanced.

**Comment by Pomax** [comment]
"JavaScript mode" is not "Processing.js". Any bugs around Pjs that aren't filed here are not going to end up being "known" in the Pjs context.

**Comment by GoToLoop** [comment]
I've lost the count on how many times you've explained the "differences" between Pjs framework library & PDE's JavaScript Mode. And again, I can't see how those picky differences would be relevant about the subject I've brought just now.

At least in Processing's camp, I've never seen n1 bringing out the subject about the differences between Processing's "PApplet" library and its corresponding PDE's Java Mode.

That is, http://ProcessingJS.org/ site is where Pjs is officially downloaded from and hosts its API online reference. Which is almost the same as "Java Mode"'s.

Pjs' main goal (now the only goal since "pure JS" was "cruelly" extirpated) is to transpile sketches written in PDE's "Java Mode" to JS syntax, so it can run under our browsers too, in addition to desktop.

Whether that transpilation is automatically accomplished by PDE's "JavaScript Mode" or directly by customizing our own ".html" file in order to load Pjs library and pass to it the Java syntax sketch somehow, those 2 approaches got the same goal at deploying ".pde" sketches to the web.

Asking again: Why http://ProcessingJS.org/ site doesn't have a link to the "JavaScript Mode" sub-category @ https://forum.Processing.org/two/categories/javascript.
How would folks know about there's a forum dedicated to Pjs once they download it?

B/c the category is called "JavaScript Mode", http://ProcessingJS.org/ shouldn't link to it is the extreme of picky! Everything about deploying sketches to run on the web under Pjs goes there, whether or not PDE's "JavaScript Mode" is actually used as an intermediary IDE for it.

Ah! I've already issued my suggestions to that to processing-js.github.io since Jan, 14th: https://github.com/processing-js/processing-js.github.io/issues/28

**Comment by Pomax** [comment]
@GoToLoop I've removed your comment, as it contributed nothing to this bug report. As always, if you want to discuss the structuce of the Processing.js project, and how it ties into Processing itself, feel free to file one or more separate issues for that. This issue is for a bug in the code for `p.noise`.
